### PR TITLE
[SNC-14] add: pagination and filtering to ListDatasets

### PR DESCRIPTION
**Pull request by thearchitector**
## Summary

Add filtering and pagination to `ListDatasets` via `DatasetFilter` and `PagedRequest`.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Checklist:

- [x] Jira story/task is put into PR title
- [x] Code has been self-reviewed
- [x] Hard-to-understand areas have been commented
- [ ] Documentation has been updated
- [x] My changes generate no new warnings
- [x] Unit tests have been added for base functionality and known edge cases
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: indico/queries/datasets.py [hunk 1]
```diff
@@ -42,21 +44,52 @@ class ListDatasets(GraphQLRequest):
...
 
-    def __init__(self, *, limit: int = 100):
-        super().__init__(self.query, variables={"limit": limit})
+    def __init__(
+        self,
+        *,
+        filters: Optional[Union[Dict, DatasetFilter]] = None,
+        limit: int = 100,
+        order_by: str = "ID",
+        desc: bool = False,
```
Comment (thearchitector) [inline indico/queries/datasets.py:80]: for backwards compat, keep limit at 100 and ascending order by id

File: tests/integration/queries/test_dataset.py [hunk 2]
```diff
@@ -78,6 +76,16 @@ def test_list_datasets(indico, airlines_dataset):
     assert type(datasets[0]) == Dataset
 
 
+def test_list_datasets_filtered(indico, airlines_dataset):
```
Comment (meghanhickey) [inline tests/integration/queries/test_dataset.py:79]: might want to test a happy path as well, just to make sure the code doesn't filter out all datasets

**Review by meghanhickey — APPROVED** [review]

File: tests/integration/queries/test_dataset.py (hunk 2 above)
Comment (meghanhickey) [reply tests/integration/queries/test_dataset.py:79]: and by "a happy path" I mean a path where you do expect some datasets to be returned

Comment (thearchitector) [reply tests/integration/queries/test_dataset.py:79]: in the same func? i was thinking the existing `test_list_datasets` test above it kinda achieves that

**Review by meghanhickey — APPROVED** [review]
ship it

**Review by meghanhickey — APPROVED** [review]

**Review by goatrocks — APPROVED** [review]

**Review by nateshim-indico — APPROVED** [review]
